src/evaluators/NewExpression.js

```diff
@@ -59,6 +62,74 @@ export default function(
     // b. ReturnIfAbrupt(argList).
   }
 
+  // If we are in pure scope, attempt to recover from creating the construct if
+  // it fails by creating a temporal abstract
+  if (realm.isInPureScope()) {
+    return tryToEvaluateConstructOrLeaveAsAbstract(constructor, argsList, strictCode, realm);
+  } else {
+    return createConstruct(constructor, argsList, realm);
+  }
+}
+
+function tryToEvaluateConstructOrLeaveAsAbstract(
+  constructor: Value,
+  argsList: Array<Value>,
+  strictCode: boolean,
+  realm: Realm
+): ObjectValue | AbstractObjectValue {
+  let effects;
+  try {
+    effects = realm.evaluateForEffects(
+      () => createConstruct(constructor, argsList, realm),
+      undefined,
+      "tryToEvaluateConstructOrLeaveAsAbstract"
+    );
+  } catch (error) {
+    // if the constructor is abstract, then create a unary op for it,
+    // otherwise we rethrow the error as we don't handle it at this
+    // point in time
+    if (error instanceof FatalError && constructor instanceof AbstractValue) {
+      // we need to havoc all the arguments
```

Comment:
You also need to havoc the constructor since it is passed to new.target or the prototype is passed as a newly created object.

Most of the time that won’t matter since the constructor is defined outside the pure function so it won’t be havoced but if it is defined inside the function it can matter.